ddht/v5/crawl.py

```diff
@@ -0,0 +1,234 @@
+
+        self.seen_nodeids: Set[NodeID] = set()
+        self.responsive_peers: Set[NodeID] = set()
+
+        self.enrqueue_send, self.enrqueue_recv = trio.open_memory_channel[ENRAPI](
+            10_000
+        )
+
+    async def fetch_enr_bucket(self, remote_enr: ENRAPI, bucket: int) -> List[ENRAPI]:
+        peer_id = remote_enr.node_id
+
+        logger.debug(f"sending FindNode. nodeid={encode_hex(peer_id)} bucket={bucket}")
+
+        try:
+            with trio.fail_after(2):
+                find_node = FindNodeMessage(
+                    request_id=self.client.message_dispatcher.get_free_request_id(
+                        peer_id
+                    ),
+                    distance=bucket,
+                )
+
+                responses = await self.client.message_dispatcher.request_nodes(
+                    peer_id, find_node
+                )
+
+                return [enr for response in responses for enr in response.message.enrs]
+        except trio.TooSlowError:
+            logger.info(
+                f"no response from peer. nodeid={encode_hex(peer_id)} bucket={bucket}"
+            )
+            raise StopScanning()
+        except UnexpectedMessage:
+            logger.exception(
+                "received a bad message from the peer. " f"nodeid={encode_hex(peer_id)}"
+            )
+            raise StopScanning()
+
+    async def scan_enr_buckets(self, remote_enr: ENRAPI) -> None:
+        """
+        Attempts to read ENRs from the outermost few buckets of the remote node.
+        """
+        peerid = remote_enr.node_id
+        logger.debug(f"scanning. nodeid={encode_hex(peerid)}")
+
+        SCAN_DEPTH = 100
+
+        consecutive_empty_buckets = 0
+
+        try:
+            for bucket in range(256, 256 - SCAN_DEPTH, -1):
+                enrs = await self.fetch_enr_bucket(remote_enr, bucket)
+
+                novel_enrs = len(
+                    [enr for enr in enrs if enr.node_id not in self.seen_nodeids]
+                )
+
+                logger.info(
+                    f"received ENRs. nodeid={encode_hex(peerid)} bucket={bucket} "
+                    f"enrs={len(enrs)} novel={novel_enrs}"
+                )
+
+                if len(enrs) == 0:
+                    consecutive_empty_buckets += 1
+                else:
+                    self.responsive_peers.add(peerid)
+                    consecutive_empty_buckets = 0
+
+                # as we get deeper into the peer's routing table the chance that there are
+                # any peers in the given bucket decreases exponentially. we save some time
+                # by moving on to the next peer once we notice that the buckets have
+                # started to think out.
+                if consecutive_empty_buckets >= 5:
+                    return
+
+                for enr in enrs:
+                    await self.schedule_enr_to_be_visited(enr)
+
+                await trio.sleep(1)  # don't overburden this peer
```

Comment:
These requests should be pretty easy for peers to respond to.  Do you think we really need this?  Did you happen to explore how this effects performance?

Reply:
I changed it to .1s and it's certainly faster, it now runs in ~5.5 minutes, I'm okay with lowering it to there